### PR TITLE
make ControlPotmeter's '_set_default' an in/out CO

### PR DESCRIPTION
**Pull request by ronso0**
inspired by the GUI lag probably caused by `<Transform><IsEqual>0.5</IsEqual></Transform>` when using a high-res pitch slider
https://mixxx.discourse.group/t/mixtrack-pro-3-tempo-slider-lag-on-w10/21835/18
``` xml
<Connection>
    <ConfigKey><Variable name="Group"/>,rate</ConfigKey>
    <Transform><IsEqual>0.5</IsEqual></Transform>
    <BindProperty>highlight</BindProperty>
</Connection>
```
~~It's not yet verified that this would fix the lag, but~~
it's handy anyway to have `group, control _set_default` with read/write capability.

**TODO**
- [x] Document new control in the manual

### CODE REVIEW

**Comment by ronso0** [comment]
@NotYourAverageAl 
Please test if the PR build fixes the GUI lag for you.

**Comment by NotYourAverageAl** [comment]
@ronso0 thanks, I can confirm this fixes the lag issue completely.

**Comment by ronso0** [comment]
okay, great!

Though, this is just a workaround.
I don't know (and am not going to investigate) if the lag you discovered with the mixtrack Pro 3 pitch fader is
* a bug with with some Qt or other depencies on windows or
* `<IsEqual>` is simply not designed/efficient for non-integer compare values.

**Comment by ronso0** [comment]
Documentation is in https://github.com/mixxxdj/manual/pull/374

**Comment by Swiftb0y** [comment]
not sure if this is bikeshedding, but I'd suggest naming the CO `rate_neutral` as it conveys the intent better and makes more sense when used in non-gui contexts like controller scripting.

**Comment by Holzhaus** [comment]
> not sure if this is bikeshedding, but I'd suggest naming the CO `rate_neutral` as it conveys the intent better and makes more sense when used in non-gui contexts like controller scripting.

or `rate_original`?

**Comment by uklotzde** [comment]
Or simply name it by what it stands for, i.e. `rate_zero` which is unambiguous and universal. We do not need to invent a new term if none exists already.

**Comment by JoergAtGithub** [comment]
Isn't a rate of zero standstill? If the slider is in the center position it's rate one.

**Comment by uklotzde** [comment]
> Isn't a rate of zero standstill? If the slider is in the center position it's rate one.

It's a relative rate.

**Comment by Holzhaus** [comment]
`rate_adjust_zero` to make it clearer? Idc ;-)

**Comment by ronso0** [comment]
thanks for the proposals, nothing really convinced me, except maybe `rate_original` -_o
but -for me- that refers to a non-scale value that cannot easily be restored, like a string or a number for example.
`rate_zero` is technically correct but we talk about 'speed' 

the new CO is a read-only **state** so if we want to get away from the slider reference it should be `rate_reset` but that is already taken by the `_reset` **action** of controlpotmeter.
`rate_is_reset` ?
`rate_restored` ?

**Comment by ronso0** [comment]
let's vote :grimacing: :grin:

**Comment by ronso0** [comment]
sry, just kidding...
I hope this has not become a 'naming issue' by now..

The more I think about it I think the original quick shot `rate_centered` makes sense, because also in non-gui (scripting) contexts COs relate to DJ hardware: basically in any DJ app and any DJ hardware the speed control is represented by a slider (except portable turntables which may have a rotary knob instead -- which can also be centered for neutral zero original rate)
I strained myself but don't see how `_centered` can be mistaken for something else.

Accordingly we also have `orientation_center` to center the crossfader.

**Comment by daschuer** [comment]
Just out of interest, can we explain the GUI lag now? Maybe we have discovered an endless loop somewhere. Something like two rounding functions that have two ideas of rounding.
To pixels vs midi values or such.

I can confirm that such a center indicator is usefull. So nothing holding back this PR.

**Comment by daschuer** [comment]
Wait ... does this apply to all sliders?
How about re-usung the "_reset" contol for it as a general solution? _reset

**Comment by Swiftb0y** [comment]
I'm sorry, I really did not intend to create a bikeshedding discussion, though my opinion on `rate_centered` vs `rate_neutral` vs `rate_original` is still strong. I prefer `rate_neutral` because just looking at the name is self-explanatory. `rate_centered` would have to be something like `rate_slider_neutral`. `rate_original` suggests some kind of timeline.

If I still didn't convince you, I'd be ok with `rate_centered` just for the sake of consistency with `orientation_center`.

**Comment by daschuer** [comment]
The existing Co is "rate_set_default"
Not an ideal name for this CO.  ... Mmm
It could become a center indicator/button.

What do you think?

**Comment by Swiftb0y** [comment]
>The existing Co is "rate_set_default"
Not an ideal name for this CO. ... Mmm

Those are COs added because `rate` is a ControlPotMeter, so the `_set_default` part is a generic suffix intended to work for all kinds of controls. `rate_centered`is just an extra CO for the `rate` independent of the ControlPotMeter.

**Comment by daschuer** [comment]
Yes, right. Currently it is only used the receive updates. It would be no deal to move the logic introduced here to this control and let all sliders an knobs benefit from it.

**Comment by ronso0** [comment]
@daschuer You're right, all potmeter controls should have this status CO.
I didn't remember we have `_set_default`. Complement would be `_is_default`.

"no big deal" may be correct if you've been working on the control system for years. I haven't, but I made it. Please test.

File: src/control/controlpotmeter.cpp
Comment (ronso0) [inline src/control/controlpotmeter.cpp:59]: first I tried to avoid the `ControlObject > defaultValue()` lookup on every value change by storing defaultValue when ControlPotmeter is initialized, but that wouldn't work for ControlPotmeter derivates that set their own default, like `ControlAudioTaperPot` "volume" for example.

> `ControlObject > defaultValue()` lookup on every value change

is that actually a performance issue?

**Comment by daschuer** [comment]
> "no big deal" may be correct

Oh sorry, this was not meant to talk the work small. Every change is a significant effort and thank you for picking it up.

**Comment by daschuer** [comment]
I am unsure about the following:

From the naming, it would be the least confusing solution to introduce the complement CO _is_default to _set_default.
When we look to the entire CO system we have push button controls that reflect the button LED state as well as receiving press and release actions from the controller. From this point of view it would be less confusing to follow it and make _set_default such a push button control. We may also add a new in/out co for this and alias/deprecate the _set_default. 

All this works for me, I just want to make sure we make an informed decision.

File: src/control/controlpotmeter.cpp
Comment (daschuer) [reply src/control/controlpotmeter.cpp:59]: Not really. I will have a look If there is an alternative.

Comment (daschuer) [reply src/control/controlpotmeter.cpp:59]: mmm, this is a performance issue if the value is changed from another thread, in this case the checkValue() is called via a queued connection. 
We can clalc the default state here and emit a signal only for the case it changes: https://github.com/mixxxdj/mixxx/blob/fa1369f880494d1e3c9aa7f41c66f2aae8212741/src/control/control.cpp#L222
This will affect "EVERY" CO though ...

Comment (daschuer) [reply src/control/controlpotmeter.cpp:59]: I think the best is to use a direct connection to ControlDoublePrivate::valueChanged() this way it is done in the thread that set the value. Like here: 
https://github.com/mixxxdj/mixxx/blob/fa1369f880494d1e3c9aa7f41c66f2aae8212741/src/control/controlobject.cpp#L50

If my thought where to confusing, I am happy to propose a PR for this change.

**Comment by Swiftb0y** [comment]
> We may also add a new in/out co for this and alias/deprecate the _set_default.

I'm in favor of this and deprecating _set_default.

**Comment by ronso0** [comment]
> We may also add a new in/out co for this and alias/deprecate the _set_default.

Sounds reasonable.

To clarify the purpose and urgency of this PR: avoid GUI lags @NotYourAverageAl encountered.
Can we merge the current solution? The rate center indicator is the first and only (GUI) control to make use of such an indicator. Something similar may also be helpful for `key` but that's working entirely different.
I'm about to finish branches more relevant  for 2.3, so any further optimization of the controls can be done in a follow-up PR by someone familiar with the contorls system.

**Comment by ronso0** [comment]
If the current implementation is okay I'll update mixxxdj/manual#374

File: src/control/controlpotmeter.cpp
Comment (ronso0) [reply src/control/controlpotmeter.cpp:59]: Not confusing but I'm a bit stingy about my spare time available for Mixxx.
Please send a PR (or push(-force) directly if you already have a working solution ready to hand.

**Comment by ronso0** [comment]
> > "no big deal" may be correct
> 
> Oh sorry, this was not meant to talk the work small. Every change is a significant effort and thank you for picking it up.

no prob, I didn't understand it that way, it's just not "swapping a few lines" only, as the perfomrmance issue indicates..

**Comment by daschuer** [comment]
https://github.com/ronso0/mixxx/pull/24 is a improved version. 
I was about to create an alias, but than I have noticed that the name _set_default is not too bad, given that it is already there. 
It suits to make a touch friendly default button and the name fits for this purpose. 

The original issue is hopefully solved with https://github.com/mixxxdj/mixxx/pull/3806

We can create an alias at any time later, while undoing the alias is hard.  ...

**Comment by ronso0** [comment]
Thanks @daschuer  !

**Comment by daschuer** [comment]
@Swiftb0y Do you want to have a final look?

File: res/skins/LateNight/decks/rate_controls_compact.xml
Comment (Swiftb0y) [inline res/skins/LateNight/decks/rate_controls_compact.xml:116]: `rate_set_default` or `rate_is_default`, what is it now?

File: res/skins/LateNight/decks/rate_controls_compact.xml
Comment (Swiftb0y) [inline res/skins/LateNight/decks/rate_controls_compact.xml:176]: same here

**Review by Swiftb0y — CHANGES_REQUESTED** [review]
Can you also document that reading from `_set_default` indicates whether the controls value is currently its default value in the manual?

File: res/skins/LateNight/decks/rate_controls_compact.xml
Comment (ronso0) [reply res/skins/LateNight/decks/rate_controls_compact.xml]: it's `rate_set_default`, https://github.com/ronso0/mixxx/pull/24 was incomplete.
fixed.

**Comment by ronso0** [comment]
As soon as all reviewers will have approved the changes I'd like to squash the c++ and and the skin changes into two commits.
The current log looks like back and forth and contains too much noise for the actual outcome.
Agreed?

**Comment by daschuer** [comment]
Yes, thank you.

**Review by Swiftb0y — APPROVED** [review]

**Comment by ronso0** [comment]
Nice, done.
Thanks everyone!

**Comment by Holzhaus** [comment]
Thanks. Please update the manual accordingly.

**Pull request merged**
